### ISSUE TITLE
Using atom-text-editor in packages

### ISSUE THREAD

**Issue by benogle** [atom/atom#5503]
For reference I've made a package to display all the standalone `atom-text-editor` permutations. https://github.com/atom/atom-text-editor-exercise

Currently using an `atom-text-editor` in a package as a replacement for `textarea` isnt that great:

```html
<atom-text-editor>
    This editor
    grows
</atom-text-editor>
```

Produces:

![screen shot 2015-02-11 at 12 10 10 pm](https://cloud.githubusercontent.com/assets/69169/6155298/f5821de6-b1e6-11e4-9e7d-0e58eb069536.png)

A few issues

* It shows line numbers by default
* it shows invisibles if I have them enabled
* It grows by default
* The styling is not like the mini editors

Constraining height makes it behave a little more like a text area, but you need to set the height in the `style` attribute or the editor does weird things:

```html
<atom-text-editor style="height: 100px">
  Some text
</atom-text-editor>
```

Produces (border added for clarity)

![screen shot 2015-02-11 at 12 14 43 pm](https://cloud.githubusercontent.com/assets/69169/6155366/98817a14-b1e7-11e4-81ce-10000d915e8f.png)

Also not great. It introduces something that is nice in the pane-item editor, but not in a textarea: you cant get close to the bottom of the editor with the cursor, there is always a 3 line padding. 

![editor-padding](https://cloud.githubusercontent.com/assets/69169/6155421/1c83812c-b1e8-11e4-91ca-8bd62c44999f.gif)

### Proposal

I propose two new tags:

* `atom-text-area`: textarea emulation. This would be an editor that
  * does not show invisibles
  * does not show the gutter
  * does not have the 3 line padding 
  * does not grow by default
  * allows the themes to style the editor as a text area
* `atom-text-input`: a replacement for the mini editors

Then we could add attributes to enable things

* `show-gutters` would show the gutter
* `auto-height` would ignore css sizing
* `read-only` would set to read only instead of the remove `tabindex` hack 

The markdown preview can just use the `atom-text-editor` to get full code support.

**Comment by nathansobo** [comment]
Do you think it's too late to invert the defaults for these options on the normal editor and make all the bells and whistles opt-in? Also, it would be great to have a better story for the editor's height behavior than requiring inline styles. I was never happy with that but it was a bit early to sort it out when I was dealing with it.

**Comment by nathansobo** [comment]
More thoughts:

* `autosize` should be `auto-height`. But really I'd like to find a better way to detect this automatically.
* `show-gutter` should actually be `show-line-numbers` since there will be multiple gutters soon.
* `readonly` should be `read-only`

What's with the 3 lines of padding?

**Comment by benogle** [comment]
> Do you think it's too late to invert the defaults for these options on the normal editor and make all the bells and whistles opt-in?

Depends on how many people are using them. I bet not many people are using these as text areas. 

> Also, it would be great to have a better story for the editor's height behavior than requiring inline styles.

Adding a new attribute allows us to give the editors emulating a textarea a default height. I think they should be fixed height with the attribute by default. Or maybe support the `rows` attribute like on textareas.

> autosize should be auto-height. But really I'd like to find a better way to detect this automatically.

I think this should be explicit. 

> show-gutter should actually be show-line-numbers since there will be multiple gutters soon.

Would we want the other gutters on these standalone editors? How would we turn the other, non-line-number gutters off?

**Comment by mark-hahn** [comment]
> Would we want the other gutters on these standalone editors?

I definitely would.  I could use it in a package I'm working on.

**Comment by nathansobo** [comment]
> Adding a new attribute allows us to give the editors emulating a textarea a default height. I think they should be fixed height with the attribute by default. Or maybe support the rows attribute like on textareas.

You've convinced me with the analogy to `textarea` elements. They don't grow with their content, right? So this would be a bonus feature.

> Would we want the other gutters on these standalone editors? How would we turn the other, non-line-number gutters off?

Yeah, perhaps `show-gutters` (plural) makes more sense for the future, but it's weird in the current context.

If we invert this we'll have to turn these all on on the editors we instantiate in panes. Could be okay though. I'd rather not have a special `standalone` property.

**Comment by benogle** [comment]
> They don't grow with their content, right?

No they dont.

> I'd rather not have a special standalone property.

I'd like to have some kind of indicator for the styles that a text editor is not a pane editor, but a textarea-like editor that should be styled like the mini editors. Maybe that could be done easily with contextual selectors, though.

**Comment by nathansobo** [comment]
> I'd like to have some kind of indicator for the styles that a text editor is not a pane editor, but a textarea-like editor that should be styled like the mini editors.

I see what you're getting at here. Are you thinking that themes would provide distinct default styling for these kinds of editors? What do you have in mind for that styling?

> Maybe that could be done easily with contextual selectors, though.

I'm not loving any of the names proposed so far, so maybe this would be worth a try. Basically we're trying to say "not in a pane". Maybe the best way to say that is contextually.

**Comment by benogle** [comment]
> Are you thinking that themes would provide distinct default styling for these kinds of editors? What do you have in mind for that styling?

Yeah. They would style them like the mini editors. They have focus indicators and they have background colors / borders that match the UI theme, not the syntax theme.

e.g. a mini editor has white background color in the light theme

![screen shot 2015-02-11 at 1 35 28 pm](https://cloud.githubusercontent.com/assets/69169/6156989/e35427fc-b1f2-11e4-8111-ca4614b981cc.png)

Then has the focus indicator when it has focus. For forms, it would be nice to have consistency. 

![screen shot 2015-02-11 at 1 34 49 pm](https://cloud.githubusercontent.com/assets/69169/6156983/d1407b9c-b1f2-11e4-8c07-a567f02b9158.png)

**Comment by benogle** [comment]
Also, in some cases, we are showing code and want the editors to be styled with the syntax colors (like in the markdown preview)

**Comment by nathansobo** [comment]
Are you down to make these changes or would you like me to do it? We need to hammer out a finalized sketch of what we plan. Maybe by editing the `Proposal` section in the first comment?

**Comment by benogle** [comment]
I can make these changes soonish, but not right now. Not super high prio. 

I still want a way to cleanly differentiate between the textarea emulating editors and editors for code. A class or an attribute would be ideal, similar to the `mini` attr. Otherwise it'll be hard to have these look/work like mini editors in some cases and like code displays in something like markdown-preview.

**Comment by nathansobo** [comment]
I'm open to an attribute, just searching for the word. Another idea is a completely different tag name like `atom-text-area`. On the `atom-text-area`, everything defaults to being disabled and can be turned on. On the `atom-text-editor`, the opposite is true. But they're essentially the same code.

**Comment by benogle** [comment]
Ooh, I like that

**Comment by benogle** [comment]
Mini editors could be `atom-text-input`s?

**Comment by nathansobo** [comment]
> Mini editors could be `atom-text-inputs`?

:+1:

**Comment by benogle** [comment]
Sweet lets do that then

**Comment by nathansobo** [comment]
I ran into difficulty this weekend trying to use the `atom-text-editor` as a standalone component in a package I'm working on for the workshop in New Zealand. In my mind, making the editor usable in an intuitive way ranks almost as high as having a good recommended view framework, so I've added the `atom` label so we can discuss it today.

My thinking has evolved a little on this. I think we should name our elements to match the names of the elements they are replacing exactly: `atom-textarea` and `atom-input`. The goal should be to make these elements *drop-in replacements* for the default HTML elements so they had the exact same behavior by default. Then, we could layer on additional behavior with custom attributes.
